src/moodle_downloader.py

```diff
@@ -0,0 +1,84 @@
+import json
+import re
+import logging
+
+import globals
+import course_retrieval
+
+
+def list_resources(args):
+    try:
+        course_name = args.course
+        if course_name == "*":
+            # List the names of all available courses
+            course_retrieval.list_courses()
+        else:
+            # List all available resources within the specified course
+            course = course_retrieval.get_course(course_name)
+            if course is not None:
+                course.list_all_resources()
+    except:
+        # TODO: add logging and log exception info (traceback) to a file
+        print("Could not list resources due to an internal error.")
+
+
+def download(args):
+    try:
+        course_name = args.course
+        resource_pattern = args.file_pattern
+        destination_path = args.destination
+
+        if destination_path is None:
+            if resource_pattern is None:
+                resource_pattern = ".*"
+            if course_name is None:
+                course_name = ".*"
+            download_via_config(course_name, resource_pattern)
+        else:
+            course = course_retrieval.get_course(course_name)
+
+            resource_names = course.get_matching_resource_names(resource_pattern)
+            for resource_name in resource_names:
+                course.download_resource(resource_name, destination_path, update_handling="replace")
+    except:
+        # TODO: add logging and log exception info (traceback) to a file
+        print("Could not download resources due to an internal error.")
+
+
+def download_via_config(req_course_name=".*", req_file_pattern=".*"):
+    print("Downloading via download config ...")
+    try:
```

Comment:
And here. I'm assuming the exception would be one of the args parameters not being specified? I think it might be better to handle that with a more explicit error message.

Reply:
Missing arguments are caught by argparse. I rather implemented these "catch alls" to ensure that a friend who studies at "Maschinenbau" and who encountered [issue 11](https://github.com/NewLordVile/tum-moodle-downloader/issues/11) has a smoother experience with the tool. But I definitely agree that more comprehensive error messages are needed.